docs/usability/creating_invocable.md

```diff
@@ -0,0 +1,685 @@
+    def GetLoggingLevel(self, loggerType):
+        ''' Get the logging level for a given type
+        base == lowest logging level supported
+        con  == Screen logging
+        txt  == plain text file logging
+        md   == markdown file logging
+        '''
+        if(loggerType == "con") and not self.Verbose:
+            return logging.WARNING
+        return logging.DEBUG
+
+    def AddCommandLineOptions(self, parser):
+        pass
+
+    def RetrieveCommandLineOptions(self, args):
+        '''  Retrieve command line options from the argparser '''
+        pass
+
+    def GetSettingsClass(self):
+        return BinaryBuildSettingsManager
+
+    def GetLoggingFileName(self, loggerType):
+        return "BINARY_BUILDLOG"
+
+    def Go(self):
+        return 0
+```
+- _GetLoggingLevel_ we can get the logging level that we care about for the type of log we are creating
+- _AddCommandLineOption_ similar to above
+- _RetrieveCommandLineOptions_ similar to above
+- _GetSettingsClass_ the class that we want to look for
+- _GetLoggingFileName_ the name of the file we want to create for txt and markdown files.
+- _Go_ is the business logic of the invocable.
+
+Now let's implement an actual method to handle main and being called from the commandline or a pip link.
+
+```python
+def main():
+    Edk2BinaryBuild().Invoke()
+
+
+if __name__ == "__main__":
+    DriverBuilder.main()  # otherwise we're in __main__ context
+```
+As you can see, we call oursevles via import rather than just directly calling main.
+This is a quirk/design flaw that might be revisited in the future, but in the meantime, this is a workaround.
+
+Now that we have a way to invoke this and execute our go, we can call if from teh commandline.
+
+If we were to run this right now, we would see this as output (assuming you created an empty settings class).
+```console
+SECTION - Init SDE
+SECTION - Loading Plugins
+SECTION - Start Invocable Tool
+SECTION - Summary
+PROGRESS - Success
+```
+
+We can see that we're initing the SDE, loading plugins and helpers, and starting the invocable. All for virtually free!
+Neato.
+
+Now let's start implementing the meat of the invokation.
+
+```python
+class Edk2BinaryBuild(Edk2Invocable):
+
+  ...
+
+    def Go(self):
+        ret = 0
+        env = shell_environment.GetBuildVars()
+        # set our environment with specific variables that we care about that EDK2 needs
+        env.SetValue("PRODUCT_NAME",
+                     self.PlatformSettings.GetName(), "Platform Hardcoded")
+        env.SetValue("BLD_*_BUILDID_STRING", "201905", "Current Version")
+        env.SetValue("BUILDREPORTING", "TRUE", "Platform Hardcoded")
+        # make sure we always do a build report
+        env.SetValue("BUILDREPORT_TYPES",
+                     'PCD DEPEX LIBRARY BUILD_FLAGS', "Platform Hardcoded")
+
+        # Run pre build hook
+        ret += self.PlatformSettings.PreFirstBuildHook()
+        # get workspace and package paths for
```

Comment:
again comment for the code.  
With this ret += pattern and you continuing to execute without checking ret it is very hard to pinpoint where the issue happens when something breaks.

Reply:
Yeah- it's fixed in basecore but wasn't fixed in the doc. I'll make it a bit more sensible